omg/cmd/get/parse.py

```diff
@@ -0,0 +1,241 @@
+
+    def generator(self):
+        """
+        Yields all resources types -> set<Resource>.
+        """
+        for r_type in self.get_types():
+            yield r_type, self.get_resources(r_type)
+
+    def __contains__(self, item):
+        """
+        Check if given resource type exists.
+        """
+        res = map_res(item)
+        if res is not None:
+            return self.has_type(res['type'])
+
+    def __iter__(self):
+        """
+        Initialize iterator from ResourceMap generator
+        """
+        self.__res_generator = self.generator()
+        return self
+
+    def __next__(self):
+        """
+        Pull next item from ResourceMap generator.
+        """
+        if self.__res_generator is None:
+            self.__iter__()
+        r_type, r_name = next(self.__res_generator)
+        return r_type, r_name
+
+    def __len__(self):
+        """
+        Length of ResourceMap.
+        """
+        return len(self.dict)
+
+
+class Method(enum.Enum):
+    """
+    Enumeration of various methods `omg get` can be called.
+    """
+    UNKNOWN = 0  # ¯\_(ツ)_/¯
+    SINGLE_TYPE = 1  # e.g. omg get pod/mypod
+    MULTI_TYPE_SINGLE_RESOURCE = 2  # e.g. omg get svc,ep,ds dns-default
+    MULTI_TYPE_MULTI_RESOURCE = 3  # e.g. omg get pod mypod myotherpod
+    ALL = 4  # e.g. oc get all
+
+
+class ResourceParseError(Exception):
+    """
+    Error raised during parsing of `oc get` args.
+    """
+    pass
+
+
+def _parse_slash(arg):
+    """
+    Parses out a single word containing a slash, validates the type is known and returns: type name, resource name
+    """
+    try:
+        o_split = arg.split('/')
+        r_type = o_split[0]
+        r_name = o_split[1]
+        checked_type = map_res(r_type)
+        if r_type is None:
+            raise ResourceParseError("[ERROR] Invalid object type: " + r_type)
+        return checked_type['type'], r_name
+    except Exception:
+        raise ResourceParseError("[ERROR] Invalid object type: " + arg)
+
+
+def parse_get_resources(objects: tuple) -> (Method, ResourceMap):
+    """
+    Takes a tuple (examples below) of arguments passed to `omg get ...` and parses it into
+    manageable class. Returns a tuple containing the Method enum and resource mapping.
+
+    Raises ResourceParseError if it runs into any issues.
+
+    Example object contents:
+    ('pod', 'httpd')
+    ('pods', 'httpd1', 'httpd2')
+    ('dc/httpd', 'pod/httpd1')
+    ('routes')
+    ('pod,svc')
+    """
+    method = Method.UNKNOWN
+    resources = ResourceMap()
+
+    if len(objects) == 0:
+        # We've nothing to parse right now.
+        return method, resources
+
+    last_object = []
+
+    # Check first arg to see if we're doing multiple or single resources
+    first = objects[0]
+    if '/' in first:
+        method = Method.MULTI_TYPE_MULTI_RESOURCE
+    elif ',' in first:
+        method = Method.MULTI_TYPE_SINGLE_RESOURCE
+    else:
+        method = Method.SINGLE_TYPE
```

Comment:
@bostrt  I agree with and like the enum approach. However the naming is a bit confusing to me (may be I am not reading it right, so feel free to correct me). I added `__str__` function in the `ResourceMap` class to make it printable and added `print` after we call `get_method, resource_list = parse.parse_get_resources(objects)` in `get_main.py`. So here are some of the outputs:

```
> omg project openshift-monitoring

> omg get svc,deployments grafana
get_method: Method.SINGLE_TYPE
resource_list: {'service': {'grafana'}, 'deployment': {'grafana'}}
NAME             TYPE       CLUSTER-IP     EXTERNAL-IP  PORT(S)   AGE
service/grafana  ClusterIP  172.30.99.136  <none>       3000/TCP  21d

NAME                READY  UP-TO-DATE  AVAILABLE  AGE
deployment/grafana  1/1    1           1          21d

>omg get svc,deployments grafana kube-state-metrics
get_method: Method.SINGLE_TYPE
resource_list: {'service': {'grafana', 'kube-state-metrics'}, 'deployment': {'grafana', 'kube-state-metrics'}}
NAME                        TYPE       CLUSTER-IP     EXTERNAL-IP  PORT(S)            AGE
service/grafana             ClusterIP  172.30.99.136  <none>       3000/TCP           21d
service/kube-state-metrics  ClusterIP  None           <none>       8443/TCP,9443/TCP  21d

NAME                           READY  UP-TO-DATE  AVAILABLE  AGE
deployment/grafana             1/1    1           1          21d
deployment/kube-state-metrics  1/1    1           1          21d

> omg get deployment/grafana deployment/kube-state-metrics
get_method: Method.MULTI_TYPE_MULTI_RESOURCE
resource_list: {'deployment': {'kube-state-metrics', 'grafana'}}
NAME                READY  UP-TO-DATE  AVAILABLE  AGE
grafana             1/1    1           1          21d
kube-state-metrics  1/1    1           1          21d
```

I think it would be more intuitive and natural to name the methods on how we are programmatically parsing e.g:

- Method.SLASH (e.g, `pod/p1` or `pod/p1 pod/p2`) 
- Method.COMMA (e.g, `pod,svc` or `pod,svc o1 o2`)
- Methond.PLAIN (e.g, `all` or `pod`, or `pod p1 p2`)

What do you think?